controller/volume_attachment_controller_test.go

```diff
@@ -1,400 +1,408 @@
+	tc.expectedVol = tc.vol.DeepCopy()
+}
+
+func (s *TestSuite) TestVolumeAttachmentLifeCycle(c *C) {
+	var tc *volumeAttachmentTestCase
+	testCases := map[string]*volumeAttachmentTestCase{}
+
+	///////////////////////////////////////////////////////////////////
+	tc = generateVolumeAttachmentTestCaseTemplate(TestVolumeName)
+	tc.volAttachment.Spec.AttachmentTickets = map[string]*longhorn.AttachmentTicket{
+		"attachment-01": &longhorn.AttachmentTicket{
+			ID:         "attachment-01",
+			Type:       longhorn.AttacherTypeCSIAttacher,
+			NodeID:     TestNode1,
+			Parameters: map[string]string{},
+			Generation: 0,
+		},
+	}
+	tc.vol.Status.OwnerID = TestNode1
+	tc.vol.Status.State = longhorn.VolumeStateDetached
+	tc.copyCurrentToExpect()
+	tc.expectedVolAttachment.Status.AttachmentTicketStatuses = map[string]*longhorn.AttachmentTicketStatus{
+		"attachment-01": &longhorn.AttachmentTicketStatus{
+			ID:        "attachment-01",
+			Satisfied: false,
+			Conditions: types.SetConditionWithoutTimestamp([]longhorn.Condition{},
+				longhorn.AttachmentStatusConditionTypeSatisfied, longhorn.ConditionStatusFalse, "", ""),
+			Generation: 0,
+		},
+	}
+	tc.expectedVol.Spec.NodeID = TestNode1
+	testCases["test case 1: attach: basic"] = tc
+	///////////////////////////////////////////////////////////////////
+
+	///////////////////////////////////////////////////////////////////
+	tc = generateVolumeAttachmentTestCaseTemplate(TestVolumeName)
+	tc.volAttachment.Spec.AttachmentTickets = map[string]*longhorn.AttachmentTicket{
+		"attachment-01": &longhorn.AttachmentTicket{
+			ID:         "attachment-01",
+			Type:       longhorn.AttacherTypeSnapshotController,
+			NodeID:     TestNode1,
+			Parameters: map[string]string{},
+			Generation: 0,
+		},
+		"attachment-02": &longhorn.AttachmentTicket{
+			ID:         "attachment-02",
+			Type:       longhorn.AttacherTypeCSIAttacher,
+			NodeID:     TestNode2,
+			Parameters: map[string]string{},
+			Generation: 0,
+		},
+	}
+	tc.vol.Status.OwnerID = TestNode1
+	tc.vol.Status.State = longhorn.VolumeStateDetached
+	tc.copyCurrentToExpect()
+	tc.expectedVolAttachment.Status.AttachmentTicketStatuses = map[string]*longhorn.AttachmentTicketStatus{
+		"attachment-01": &longhorn.AttachmentTicketStatus{
+			ID:        "attachment-01",
+			Satisfied: false,
+			Conditions: types.SetConditionWithoutTimestamp([]longhorn.Condition{},
+				longhorn.AttachmentStatusConditionTypeSatisfied, longhorn.ConditionStatusFalse, "", ""),
+			Generation: 0,
+		},
+		"attachment-02": &longhorn.AttachmentTicketStatus{
+			ID:        "attachment-02",
+			Satisfied: false,
+			Conditions: types.SetConditionWithoutTimestamp([]longhorn.Condition{},
+				longhorn.AttachmentStatusConditionTypeSatisfied, longhorn.ConditionStatusFalse, "", ""),
+			Generation: 0,
+		},
+	}
+	tc.expectedVol.Spec.NodeID = TestNode2
+	testCases["test case 2: attach: multiple attachments"] = tc
+	///////////////////////////////////////////////////////////////////
+
```

Comment:
When you take out the comments, this is the same set of tests with Attachment replaced by AttachmentTicketStatus.  The field mappings look reasonable.
Tests 2 and 3 formerly had error message text to check, and that's not possible any more?  (Test 5 has a ReasonAttached string, so it must be possible sometimes.)

Reply:
Since the last time I wrote this unit test, the AD controller have been evolve quite a bit so the error message and struct field need to be modified.

> Tests 2 and 3 formerly had error message text to check, and that's not possible any more?

Yeah, you are correct

> Test 5 has a ReasonAttached string, so it must be possible sometimes.

This is also correct